src/simulators/density_matrix/densitymatrix_thrust.hpp

```diff
@@ -454,84 +459,217 @@ double DensityMatrixThrust<data_t>::probability(const uint_t outcome) const {
 }
 
 template <typename data_t>
-reg_t DensityMatrixThrust<data_t>::sample_measure(const std::vector<double> &rnds) const {
+class DensityMatrixThrust_Gather
+{
+protected:
+  uint_t offset;
+  data_t* pVec;
+public:
+  DensityMatrixThrust_Gather(data_t* pVecIn,uint_t offsetIn)
+  {
+    pVec = pVecIn;
+    offset = (offsetIn + 1) * 2;
+  }
+
+  __host__ __device__ void operator()(uint_t iter)
+  {
+    uint_t idx;
+    idx = iter * offset;
+    pVec[iter] = pVec[idx];
+  }
+};
 
-  const int_t END = 1LL << num_qubits();
+
+
+template <typename data_t>
+reg_t DensityMatrixThrust<data_t>::sample_measure(const std::vector<double> &rnds) const 
+{
   const int_t SHOTS = rnds.size();
-  reg_t samples;
+  reg_t samples,localSamples;
+  std::vector<double> placeSum;
+  data_t* pVec;
+  uint_t i;
+  double sum,localSum,globalSum;
+  int iPlace;
+  reg_t diagIdx(BaseVector::m_nPlaces+1);
+
+#ifdef AER_TIMING
+  TimeStart(QS_GATE_MEASURE);
+#endif
+
+  BaseVector::UpdateReferencedValue();
```

Comment:
This is required to handle &operator[] because we can not directory put state value on GPU by using operator[] from outside of qubitvector class. 
So we temporary put the value put from outside in the host memory and UpdateReferencedValue writes back to GPU.